cpp/src/arrow/flight/test_definitions.cc

```diff
@@ -518,22 +589,49 @@ void DataTest::TestIssue5095() {
 //------------------------------------------------------------
 // Specific tests for DoPut
 
+static constexpr char kExpectedMetadata[] = "foo bar";
+
 class DoPutTestServer : public FlightServerBase {
  public:
   Status DoPut(const ServerCallContext& context,
                std::unique_ptr<FlightMessageReader> reader,
                std::unique_ptr<FlightMetadataWriter> writer) override {
     descriptor_ = reader->descriptor();
+
+    if (descriptor_.type == FlightDescriptor::DescriptorType::CMD) {
+      if (descriptor_.cmd == "TestUndrained") {
+        // Don't read all the messages
+        return Status::OK();
+      }
+    }
+
     int counter = 0;
+    FlightStreamChunk chunk;
     while (true) {
-      FlightStreamChunk chunk;
       RETURN_NOT_OK(reader->Next(&chunk));
       if (!chunk.data) break;
+      if (counter % 2 == 1) {
```

Comment:
What happens if `counter % 2 == 0`? Should we assert that there's no app metadata?